webserver/webserver/src/test/java/io/helidon/webserver/HttpPipelineTest.java

```diff
@@ -0,0 +1,204 @@
+                            executor.schedule(() -> res.status(Http.Status.OK_200).send("Response " + n + "\n"),
+                                    delay, TimeUnit.MILLISECONDS);
+                        })
+                        .build())
+                .build()
+                .start()
+                .toCompletableFuture()
+                .get(10, TimeUnit.SECONDS);
+
+        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder()
+                .addNetworkInterceptor(new LoggingInterceptor())
+                .protocols(Collections.singletonList(Protocol.H2_PRIOR_KNOWLEDGE));
+        client = clientBuilder.build();
+
+        LOGGER.info("Started server at: https://localhost:" + webServer.port());
+    }
+
+    /**
+     * Pipelines request_0 and request_1 and makes sure responses are returned in the
+     * correct order. Note that the server will delay the response for request_0 to
+     * make sure they are properly synchronized.
+     *
+     * @throws Exception If there are connection problems.
+     */
+    @Test
+    public void testPipelining() throws Exception {
+        try (SocketHttpClient s = new SocketHttpClient(webServer)) {
+            s.request(Http.Method.PUT, "/");        // reset server
+            s.request(Http.Method.GET, "/");        // request_0
+            s.request(Http.Method.GET, "/");        // request_1
+            String put = s.receive();
+            assertThat(put, notNullValue());
+            String get0 = s.receive();
+            assertThat(get0, containsString("Response 0"));
+            String get1 = s.receive();
+            assertThat(get1, containsString("Response 1"));
+        }
+    }
+
+    /**
+     * Same as previous test but using HTTP/2 and OkHttp as async client.
+     */
+    @Test
+    public void testPipeliningHttp2() throws Exception {
+        MediaType mt = MediaType.get("text/plain");
+        URL url = new URL("http://localhost:" + webServer.port() + "/");
+        Request put = new Request.Builder().url(url).put(RequestBody.create(mt, "")).build();
+        client.newCall(put).execute();
+
+        Request get = new Request.Builder().url(url).build();
+        CompletableFuture<?> cf0 = new CompletableFuture<>();
+        CompletableFuture<?> cf1 = new CompletableFuture<>();
+
+        client.newCall(get).enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException e) {
+                cf0.completeExceptionally(e);
+            }
+
+            @Override
+            public void onResponse(Call call, Response r0) throws IOException {
+                LOGGER.info("Received r0");
+                if (cf1.isDone()) {
+                    LOGGER.info("Expected r0 before r1");
+                    cf0.completeExceptionally(new RuntimeException("Expected r0 before r1"));
+                } else {
+                    assertThat(r0.body().string(), containsString("Response 0"));
```

Comment:
Will be the assertion error thrown on correct thread?

Reply:
No, will fix this.